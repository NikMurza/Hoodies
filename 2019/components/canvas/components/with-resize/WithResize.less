@import (reference) '~@common/styles/constants/colors.less';

@resize-marker-padding: 4px;

.scale-resize-marker(@deg: 0) {
  transform: rotate(@deg);
  &.hover {
    transform: rotate(@deg) scale(1.5);
  }
  &.hide {
    transform: rotate(@deg) scale(0);
  }
}

.with-resize {
  position: absolute;
  height: inherit;

  .component {
    height: inherit;
    &.resizing {
      pointer-events: none;
    }
  }
}

.with-move {
  .with-resize {
    position: static;
  }
}

// цвет области при ресайзе будет fade(@blue, 8%)
.with-resize__selection-area {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  border: 2px solid @blue;
  border-radius: 2px;
  z-index: 52;
  transition: background-color 0.3s ease-out;
  clip-path: polygon(
    -1% -1%,
    -1% 101%,
    2% 101%,
    2% 2%,
    98% 2%,
    98% 98%,
    2% 98%,
    2% 105%,
    105% 105%,
    105% -1%
  );

  &.invalid {
    border-color: @red;

    .resize-marker {
      color: @red;
    }

    .size-lbl {
      background-color: @red;
    }
  }

  &.resizing {
    z-index: 51;
    background: #3b5edc33;
    clip-path: none;
  }

  &.hover {
    .resize-marker {
      opacity: 1;
    }
  }

  &.fetching {
    .resize-marker {
      display: none;
    }
  }

  .size-lbl {
    position: absolute;
    bottom: -27px;
    left: 50%;
    transform: translateX(-50%);
    padding: 2px 4px;
    color: @white;
    background: @blue;
    border-radius: 2px;
    font-size: 11px;
    line-height: 13px;
    letter-spacing: 0.004em;
    opacity: 0;
    transition: opacity 0.3s ease-out;

    &.visible {
      opacity: 1;
    }
  }

  .resize-marker {
    position: absolute;
    width: 16px;
    color: @blue;
    transition: all 0.15s ease-out;
    box-sizing: content-box;
    padding: @resize-marker-padding;
    opacity: 0;
    svg {
      position: absolute;
    }
  }
  .resize-marker-corner {
    height: 16px;
    transform-origin: 1px 1px;
  }
  .resize-marker-center {
    display: flex;
    height: 4px;
  }
  .resize-marker-top-left {
    top: -3px - @resize-marker-padding;
    left: -3px - @resize-marker-padding;
    cursor: nwse-resize;
    .scale-resize-marker();
    &.hover {
      top: -3px - @resize-marker-padding*1.5;
      left: -3px - @resize-marker-padding*1.5;
    }
  }
  .resize-marker-top-right {
    top: -3px - @resize-marker-padding;
    right: -17px - @resize-marker-padding*3;
    cursor: nesw-resize;
    .scale-resize-marker(90deg);
    &.hover {
      top: -3px - @resize-marker-padding*1.5;
      right: -17px - @resize-marker-padding*3.5;
    }
  }
  .resize-marker-bottom-left {
    bottom: -17px - @resize-marker-padding*3;
    left: -3px - @resize-marker-padding;
    cursor: nesw-resize;
    .scale-resize-marker(-90deg);
    &.hover {
      bottom: -17px - @resize-marker-padding*3.5;
      left: -3px - @resize-marker-padding*1.5;
    }
  }
  .resize-marker-bottom-right {
    bottom: -17px - @resize-marker-padding*3;
    right: -17px - @resize-marker-padding*3;
    cursor: nwse-resize;
    .scale-resize-marker(180deg);
    &.hover {
      bottom: -17px - @resize-marker-padding*3.5;
      right: -17px - @resize-marker-padding*3.5;
    }
  }
  .resize-marker-center-top {
    top: -3px - @resize-marker-padding;
    left: calc(50% - 8px - @resize-marker-padding);
    cursor: ns-resize;
    .scale-resize-marker();
  }
  .resize-marker-center-bottom {
    bottom: -3px - @resize-marker-padding;
    left: calc(50% - 8px - @resize-marker-padding);
    cursor: ns-resize;
    .scale-resize-marker();
  }
  .resize-marker-center-left {
    top: 50%;
    left: -9px - @resize-marker-padding;
    cursor: ew-resize;
    .scale-resize-marker(90deg);
  }
  .resize-marker-center-right {
    top: 50%;
    right: -9px - @resize-marker-padding;
    cursor: ew-resize;
    .scale-resize-marker(90deg);
  }
}
