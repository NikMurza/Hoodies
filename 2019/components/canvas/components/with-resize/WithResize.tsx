import * as React from 'react';
import {
  VisualEditorRectPoints,
  VisualEditorActions,
  VisualEditorIdentity,
  VisualEditorPosition,
  VisualEditorSize,
} from '@common/components/lib/types/visual-editor';
import classNames from 'classnames';
import SVGIcon from '@common/components/lib/controls/svg-icon/SVGIcon';
import { getIconPath } from '@common/api/utils/object';
import './WithResize.less';
import { VisualizerPositionOptions } from '@common/api/types/visualizers/root';

enum ResizeMarkerType {
  Corner = 'corner',
  Center = 'center',
}

export enum ResizeMarkerPosition {
  TopLeft = 'top-left',
  TopRight = 'top-right',
  BottomLeft = 'bottom-left',
  BottomRight = 'bottom-right',
  CenterTop = 'center-top',
  CenterBottom = 'center-bottom',
  CenterLeft = 'center-left',
  CenterRight = 'center-right',
}

const resizeMarkers = [
  {
    type: ResizeMarkerType.Corner,
    position: ResizeMarkerPosition.TopLeft,
  },
  {
    type: ResizeMarkerType.Corner,
    position: ResizeMarkerPosition.TopRight,
  },
  {
    type: ResizeMarkerType.Corner,
    position: ResizeMarkerPosition.BottomLeft,
  },
  {
    type: ResizeMarkerType.Corner,
    position: ResizeMarkerPosition.BottomRight,
  },
  {
    type: ResizeMarkerType.Center,
    position: ResizeMarkerPosition.CenterTop,
  },
  {
    type: ResizeMarkerType.Center,
    position: ResizeMarkerPosition.CenterBottom,
  },
  {
    type: ResizeMarkerType.Center,
    position: ResizeMarkerPosition.CenterLeft,
  },
  {
    type: ResizeMarkerType.Center,
    position: ResizeMarkerPosition.CenterRight,
  },
];

export const defaultMinSize = { width: 200, height: 200 };
export const defaultMaxSize = { width: 800, height: 800 };

export const useEventListeners = (onMouseMove, onMouseUp, markerDragging, resizePoints): void => {
  React.useEffect(() => {
    if (markerDragging) {
      document.addEventListener('mousemove', onMouseMove);
    }

    return (): void => {
      document.removeEventListener('mousemove', onMouseMove);
    };
  }, [markerDragging]);

  React.useEffect(() => {
    if (markerDragging) {
      document.addEventListener('mouseup', onMouseUp);
    }

    return (): void => {
      document.removeEventListener('mouseup', onMouseUp);
    };
  }, [markerDragging, resizePoints]);
};

/**
 * Делаем все стороны одинакового размера
 */
export const setEqualSize = (
  points: VisualEditorRectPoints,
  width: number,
  height: number,
  markerDragging: ResizeMarkerPosition,
): void => {
  const { top, right, bottom, left } = points;

  const newWidth = width - left - right;
  const newHeight = height - top - bottom;

  switch (markerDragging) {
    case ResizeMarkerPosition.TopLeft:
      if (newWidth > newHeight) {
        points.top = -(newWidth - height);
      } else {
        points.left = -(newHeight - width);
      }
      break;
    case ResizeMarkerPosition.TopRight:
      if (newWidth > newHeight) {
        points.top = -(newWidth - height);
      } else {
        points.right = -(newHeight - width);
      }
      break;
    case ResizeMarkerPosition.BottomLeft:
      if (newWidth > newHeight) {
        points.bottom = -(newWidth - height);
      } else {
        points.left = -(newHeight - width);
      }
      break;
    case ResizeMarkerPosition.BottomRight:
      if (newWidth > newHeight) {
        points.bottom = -(newWidth - height);
      } else {
        points.right = -(newHeight - width);
      }
      break;
    default:
  }
};

/**
 * Проверяем что размер не больше и не меньше заданного в minSize / maxSize
 */
export const checkResizePointsAndSet = (
  points: VisualEditorRectPoints,
  shiftKey: boolean,
  props: {
    width: number;
    height: number;
    minSize: VisualEditorSize;
    maxSize: VisualEditorSize;
    markerDragging: ResizeMarkerPosition;
    setResizePoints: (points: VisualEditorRectPoints) => void; // TODO type
  },
): void => {
  const { top, right, bottom, left } = points;
  const { width, height, minSize, maxSize, markerDragging, setResizePoints } = props;

  const newWidth = width - left - right;
  const newHeight = height - top - bottom;

  if (newWidth < minSize.width) {
    if (Math.abs(left) > Math.abs(right)) {
      points.left = width - minSize.width - right;
    } else {
      points.right = width - minSize.width - left;
    }
  }
  if (newHeight < minSize.height) {
    if (Math.abs(top) > Math.abs(bottom)) {
      points.top = height - minSize.height - bottom;
    } else {
      points.bottom = height - minSize.height - top;
    }
  }

  if (newWidth > maxSize.width) {
    if (Math.abs(left) > Math.abs(right)) {
      points.left = width - maxSize.width - right;
    } else {
      points.right = width - maxSize.width - left;
    }
  }
  if (newHeight > maxSize.height) {
    if (Math.abs(top) > Math.abs(bottom)) {
      points.top = height - maxSize.height - bottom;
    } else {
      points.bottom = height - maxSize.height - top;
    }
  }

  if (shiftKey) {
    setEqualSize(points, width, height, markerDragging);
  }
  setResizePoints(points);
};
export interface Props extends VisualEditorIdentity, VisualEditorPosition, VisualEditorSize {
  minSize?: VisualEditorSize;
  maxSize?: VisualEditorSize;
  isSelected: boolean;
  isValid: boolean;
  isFetching?: boolean;
  isEditing?: boolean;
  onAction?(
    action: VisualEditorActions,
    data: VisualEditorIdentity & VisualEditorPosition & VisualEditorSize,
  ): void;
  zoom?: number;
}

const WithResize = <P extends object>(
  WrappedComponent: React.ComponentType<P>,
): React.FC<P & Props> => (props: Props): JSX.Element => {
  const {
    id,
    parentId,
    width,
    height,
    x,
    y,
    isSelected,
    isFetching,
    isValid,
    isEditing,
    minSize = defaultMinSize,
    maxSize = defaultMaxSize,
    onAction,
    zoom,
  } = props;

  const [isHovered, setIsHovered] = React.useState<boolean>(false);
  const [markerHovered, setMarkerHovered] = React.useState<ResizeMarkerPosition>(null);
  const [markerDragging, setMarkerDragging] = React.useState<ResizeMarkerPosition>(null);
  // запоминаем начальную позицию курсора
  const [initialCursorPosition, setInitialCursorPosition] = React.useState<VisualEditorPosition>(
    null,
  );
  // точки для области ресайза
  const [resizePoints, setResizePoints] = React.useState<VisualEditorRectPoints>(null);
  // состояние нажатия на ЛКМ
  const [isMouseButtonPressed, setIsMouseButtonPressed] = React.useState<boolean>(false);

  const onMouseEnter = React.useCallback(
    (marker: { type: ResizeMarkerType; position: ResizeMarkerPosition }) => (): void => {
      setMarkerHovered(marker.position);
    },
    [],
  );

  const onMouseLeave = React.useCallback((): void => {
    setMarkerHovered(null);
  }, []);

  const onMouseDown = React.useCallback(
    (marker: { type: ResizeMarkerType; position: ResizeMarkerPosition }) => (
      event: React.MouseEvent<HTMLDivElement, MouseEvent>,
    ): void => {
      event.preventDefault();
      event.stopPropagation();
      event.nativeEvent.stopImmediatePropagation();

      setMarkerDragging(marker.position);
      setInitialCursorPosition({ x: event.pageX, y: event.pageY });
    },
    [],
  );

  const getResizeOptions = React.useCallback((): VisualizerPositionOptions => {
    if (!resizePoints) return null;

    const { top, right, bottom, left } = resizePoints;

    return {
      id,
      x: x + left,
      y: y + top,
      width: width - left - right,
      height: height - top - bottom,
      parentId,
    };
  }, [resizePoints]);

  const onMouseUp = React.useCallback((): void => {
    if (markerDragging && resizePoints) {
      const resizeOpts = getResizeOptions();
      resizeOpts && onAction && onAction(VisualEditorActions.Resize, resizeOpts);
      setMarkerDragging(null);
      setInitialCursorPosition(null);
      setResizePoints({ top: 0, right: 0, bottom: 0, left: 0 });
      setIsMouseButtonPressed(false);
    }
  }, [markerDragging, resizePoints]);

  // Был замечен баг с несрабатыванием mouseUp в addEventListener,
  // из-за чего маркер ресайза продолжал двигаться за мышкой.
  // Чтобы удостовериться, что resizePoints сброисились, вызываем onMouseUp,
  // если ЛКМ оказалась незажатой
  React.useEffect(() => {
    if (!isMouseButtonPressed) {
      onMouseUp();
    }
  }, [isMouseButtonPressed]);

  const onMouseMove = React.useCallback(
    (event: MouseEvent): void => {
      setIsMouseButtonPressed(!!event.buttons);

      if (markerDragging) {
        const { x, y } = initialCursorPosition;
        const offsetX = (x - event.pageX) / zoom;
        const offsetY = (y - event.pageY) / zoom;

        const props = {
          width,
          height,
          minSize,
          maxSize,
          markerDragging,
          setResizePoints,
        };
        const checkResizePointsAndSetWrapper = (points: VisualEditorRectPoints): void => {
          checkResizePointsAndSet(points, event.shiftKey, props);
        };

        // в зависимости от маркера - меняем размеры области ресайза
        switch (markerDragging) {
          case ResizeMarkerPosition.TopLeft:
            checkResizePointsAndSetWrapper({ top: -offsetY, right: 0, bottom: 0, left: -offsetX });
            break;
          case ResizeMarkerPosition.TopRight:
            checkResizePointsAndSetWrapper({ top: -offsetY, right: offsetX, bottom: 0, left: 0 });
            break;
          case ResizeMarkerPosition.BottomLeft:
            checkResizePointsAndSetWrapper({ top: 0, right: 0, bottom: offsetY, left: -offsetX });
            break;
          case ResizeMarkerPosition.BottomRight:
            checkResizePointsAndSetWrapper({ top: 0, right: offsetX, bottom: offsetY, left: 0 });
            break;

          case ResizeMarkerPosition.CenterTop:
            checkResizePointsAndSetWrapper({ top: -offsetY, right: 0, bottom: 0, left: 0 });
            break;
          case ResizeMarkerPosition.CenterRight:
            checkResizePointsAndSetWrapper({ top: 0, right: offsetX, bottom: 0, left: 0 });
            break;
          case ResizeMarkerPosition.CenterBottom:
            checkResizePointsAndSetWrapper({ top: 0, right: 0, bottom: offsetY, left: 0 });
            break;
          case ResizeMarkerPosition.CenterLeft:
            checkResizePointsAndSetWrapper({ top: 0, right: 0, bottom: 0, left: -offsetX });
            break;
          default:
        }
      }
    },
    [markerDragging, initialCursorPosition, checkResizePointsAndSet],
  );

  useEventListeners(onMouseMove, onMouseUp, markerDragging, resizePoints);

  const onMouseOver = React.useCallback((e): void => {
    if (!e.target.closest('.popup-menu')) setIsHovered(true);
  }, []);

  const onMouseOut = React.useCallback((): void => {
    setIsHovered(false);
  }, []);

  const renderMarker = React.useCallback(
    (item: { type: ResizeMarkerType; position: ResizeMarkerPosition }): JSX.Element => {
      // увеличиваем в размере маркер, на котором курсор
      const isHovered = markerHovered === item.position;
      // прячем все остальные
      const hideMarker = markerHovered && !isHovered;

      return (
        <div
          onMouseEnter={onMouseEnter(item)}
          onMouseLeave={onMouseLeave}
          onMouseDown={onMouseDown(item)}
          key={`${item.type}_${item.position}`}
          className={classNames(
            `resize-marker resize-marker-${item.type} resize-marker-${item.position}`,
            {
              hide: hideMarker,
              hover: isHovered,
            },
          )}
        >
          <SVGIcon src={getIconPath(`resize-marker-${item.type}`)} />
        </div>
      );
    },
    [markerHovered, onMouseEnter, onMouseLeave, onMouseDown, zoom],
  );

  const { top = 0, bottom = 0, left = 0, right = 0 } = resizePoints || {};

  const resizeStyle = markerDragging
    ? {
        top: top * zoom,
        bottom: bottom * zoom,
        left: left * zoom,
        right: right * zoom,
      }
    : {};

  const resize = getResizeOptions();

  return (
    <div
      className={classNames('with-resize')}
      style={{ left: x, top: y }}
      onMouseOver={onMouseOver}
      onMouseOut={onMouseOut}
    >
      {isSelected ? (
        <div
          className={classNames('with-resize__selection-area', {
            resizing: !!markerDragging,
            invalid: !isValid,
            hover: isHovered && !isEditing,
            fetching: isFetching,
          })}
          style={resizeStyle}
        >
          <div
            className={classNames('size-lbl', {
              visible: resize && !!markerDragging,
            })}
          >
            {Math.round(resize?.width)} x {Math.round(resize?.height)}
          </div>
          {resizeMarkers.map(renderMarker)}
        </div>
      ) : null}
      <div
        className={classNames('component', {
          resizing: !!markerDragging,
        })}
      >
        <WrappedComponent {...(props as P)} isReadyToResize={!!markerDragging} />
      </div>
    </div>
  );
};

export default WithResize;
