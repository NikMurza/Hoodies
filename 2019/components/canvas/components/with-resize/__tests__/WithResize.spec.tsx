import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { renderHook, act } from '@testing-library/react-hooks';
import WithResize, {
  ResizeMarkerPosition,
  defaultMinSize,
  defaultMaxSize,
  useEventListeners,
  setEqualSize,
  checkResizePointsAndSet,
} from '../WithResize';

const HOC = WithResize(() => <div>Component</div>);

describe('WithResize HOC', () => {
  const props = {
    id: '1',
    width: 640,
    height: 560,
    x: 100,
    y: 200,
    isSelected: false,
    isValid: true,
    minSize: defaultMinSize,
    maxSize: defaultMaxSize,
    onAction: jest.fn(),
    isEditing: false,
  };
  const wrapper = shallow(<HOC {...props} />);
  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly if selected', () => {
    const localProps = {
      ...props,
      isSelected: true,
    };
    const wrapper = shallow(<HOC {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('onMouseEnter/Leave', () => {
    const localProps = {
      ...props,
      isSelected: true,
    };
    const wrapper = shallow(<HOC {...localProps} />);
    const topLeft = wrapper.find('.resize-marker-top-left');
    topLeft.simulate('mouseenter');
    const topLeftUpdated = wrapper.find('.resize-marker-top-left');
    const hidden = wrapper.find('.hide');
    expect(topLeftUpdated.hasClass('hover')).toBeTruthy();
    expect(hidden).toHaveLength(7);

    topLeft.simulate('mouseleave');
    const topLeftUpdated2 = wrapper.find('.resize-marker-top-left');
    const hiddenUpdated = wrapper.find('.hide');
    expect(topLeftUpdated2.hasClass('hover')).toBeFalsy();
    expect(hiddenUpdated).toHaveLength(0);
  });

  it('onMouseDown', () => {
    const localProps = {
      ...props,
      isSelected: true,
    };
    const wrapper = shallow(<HOC {...localProps} />);
    const topLeft = wrapper.find('.resize-marker-top-left');
    topLeft.simulate('mousedown', {
      preventDefault: jest.fn,
      stopPropagation: jest.fn,
      nativeEvent: { stopImmediatePropagation: jest.fn },
    });
    const resizeArea = wrapper.find('.with-resize__selection-area');
    const component = wrapper.find('.component');
    expect(resizeArea.hasClass('resizing')).toBeTruthy();
    expect(component.hasClass('resizing')).toBeTruthy();
  });

  describe('setEqualSize', () => {
    it('setEqualSize - TopLeft, newWidth > newHeight', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      setEqualSize(resizePoints, 500, 400, ResizeMarkerPosition.TopLeft);
      expect(resizePoints).toEqual({ top: -80, right: 10, bottom: 10, left: 10 });
    });

    it('setEqualSize - TopLeft, newWidth <= newHeight', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      setEqualSize(resizePoints, 400, 500, ResizeMarkerPosition.TopLeft);
      expect(resizePoints).toEqual({ top: 10, right: 10, bottom: 10, left: -80 });
    });

    it('setEqualSize - BottomRight, newWidth > newHeight', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      setEqualSize(resizePoints, 500, 400, ResizeMarkerPosition.BottomRight);
      expect(resizePoints).toEqual({ top: 10, right: 10, bottom: -80, left: 10 });
    });

    it('setEqualSize - BottomRight, newWidth <= newHeight', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      setEqualSize(resizePoints, 400, 500, ResizeMarkerPosition.BottomRight);
      expect(resizePoints).toEqual({ top: 10, right: -80, bottom: 10, left: 10 });
    });
  });

  describe('checkResizePointsAndSet', () => {
    const defaultProps = {
      width: 400,
      height: 400,
      minSize: defaultMinSize,
      maxSize: defaultMaxSize,
      markerDragging: ResizeMarkerPosition.TopLeft,
      setResizePoints: jest.fn(),
    };
    it('checkResizePointsAndSet - common case', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      const shiftKey = false;
      checkResizePointsAndSet(resizePoints, shiftKey, defaultProps);
      expect(resizePoints).toEqual({ top: 10, right: 10, bottom: 10, left: 10 });
    });

    it('checkResizePointsAndSet - common case, shiftKey', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      const shiftKey = true;
      const props = {
        ...defaultProps,
        width: 500,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: -80, right: 10, bottom: 10, left: 10 });
    });

    it('checkResizePointsAndSet - newWidth < minWidth && |left| <= |right|', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        width: 100,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: 10, right: -110, bottom: 10, left: 10 });
    });

    it('checkResizePointsAndSet - newWidth < minWidth && |left| > |right|', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 15 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        width: 100,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: 10, right: 10, bottom: 10, left: -110 });
    });

    it('checkResizePointsAndSet - newHeight < minHeight && |top| <= |bottom|', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        height: 100,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: 10, right: 10, bottom: -110, left: 10 });
    });

    it('checkResizePointsAndSet - newHeight < minHeight && |top| > |bottom|', () => {
      const resizePoints = { top: 15, right: 10, bottom: 10, left: 10 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        height: 100,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: -110, right: 10, bottom: 10, left: 10 });
    });

    it('checkResizePointsAndSet - newWidth > maxWidth && |left| <= |right|', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        width: 900,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: 10, right: 90, bottom: 10, left: 10 });
    });

    it('checkResizePointsAndSet - newWidth > maxWidth && |left| > |right|', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 15 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        width: 900,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: 10, right: 10, bottom: 10, left: 90 });
    });

    it('checkResizePointsAndSet - newHeight > maxHeight && |top| <= |bottom|', () => {
      const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        height: 900,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: 10, right: 10, bottom: 90, left: 10 });
    });

    it('checkResizePointsAndSet - newHeight > maxHeight && |top| > |bottom|', () => {
      const resizePoints = { top: 15, right: 10, bottom: 10, left: 10 };
      const shiftKey = false;
      const props = {
        ...defaultProps,
        height: 900,
      };
      checkResizePointsAndSet(resizePoints, shiftKey, props);
      expect(resizePoints).toEqual({ top: 90, right: 10, bottom: 10, left: 10 });
    });
  });

  describe('event listeners', () => {
    document.addEventListener = jest.fn();
    document.removeEventListener = jest.fn();

    const onMouseMove = jest.fn();
    const onMouseUp = jest.fn();
    const markerDragging = ResizeMarkerPosition.TopLeft;
    const resizePoints = { top: 10, right: 10, bottom: 10, left: 10 };

    beforeEach(() => {
      (document.addEventListener as jest.Mock).mockClear();
      (document.removeEventListener as jest.Mock).mockClear();
      onMouseMove.mockClear();
      onMouseUp.mockClear();
    });

    it('call listeners', () => {
      renderHook(() => useEventListeners(onMouseMove, onMouseUp, markerDragging, resizePoints));
      expect(document.addEventListener).toBeCalledTimes(2);
    });
  });
});
