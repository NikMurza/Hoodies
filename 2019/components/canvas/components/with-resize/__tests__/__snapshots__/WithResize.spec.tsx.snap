// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithResize HOC renders correctly 1`] = `
<div
  className="with-resize"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "left": 100,
      "top": 200,
    }
  }
>
  <div
    className="component"
  >
    <Component
      height={560}
      id="1"
      isEditing={false}
      isReadyToResize={false}
      isSelected={false}
      isValid={true}
      maxSize={
        Object {
          "height": 800,
          "width": 800,
        }
      }
      minSize={
        Object {
          "height": 200,
          "width": 200,
        }
      }
      onAction={[MockFunction]}
      width={640}
      x={100}
      y={200}
    />
  </div>
</div>
`;

exports[`WithResize HOC renders correctly if selected 1`] = `
<div
  className="with-resize"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "left": 100,
      "top": 200,
    }
  }
>
  <div
    className="with-resize__selection-area"
    style={Object {}}
  >
    <div
      className="size-lbl"
    >
      NaN
       x 
      NaN
    </div>
    <div
      className="resize-marker resize-marker-corner resize-marker-top-left"
      key="corner_top-left"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
    <div
      className="resize-marker resize-marker-corner resize-marker-top-right"
      key="corner_top-right"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
    <div
      className="resize-marker resize-marker-corner resize-marker-bottom-left"
      key="corner_bottom-left"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
    <div
      className="resize-marker resize-marker-corner resize-marker-bottom-right"
      key="corner_bottom-right"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
    <div
      className="resize-marker resize-marker-center resize-marker-center-top"
      key="center_center-top"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
    <div
      className="resize-marker resize-marker-center resize-marker-center-bottom"
      key="center_center-bottom"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
    <div
      className="resize-marker resize-marker-center resize-marker-center-left"
      key="center_center-left"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
    <div
      className="resize-marker resize-marker-center resize-marker-center-right"
      key="center_center-right"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </div>
  </div>
  <div
    className="component"
  >
    <Component
      height={560}
      id="1"
      isEditing={false}
      isReadyToResize={false}
      isSelected={true}
      isValid={true}
      maxSize={
        Object {
          "height": 800,
          "width": 800,
        }
      }
      minSize={
        Object {
          "height": 200,
          "width": 200,
        }
      }
      onAction={[MockFunction]}
      width={640}
      x={100}
      y={200}
    />
  </div>
</div>
`;
